### ISSUE TITLE
onMetaData: videocodecid not defined for HEVC, VP9 and AV1

### ISSUE THREAD

**Issue by ThibaultBee** [veovera/enhanced-rtmp#8]
Hi,

I am really glad a new version of RTMP is coming 🙏 

In the FLV spec, there is a field in `onMetaData` called `videocodecid` where the codec was specified:
```
videocodecid: Number: Video codec ID used in the file (see E.4.3.1 for available CodecID values)
```

with CodecID:
```
Codec Identifier. The following values are defined: 2 = Sorenson H.263
3 = Screen video
4 = On2 VP6
5 = On2 VP6 with alpha channel 6 = Screen video version 2
7 = AVC
```

Unfortunately, in the current version of enhancements of RTMP spec, there is no mention of `onMetaData`'s `videocodecid` field.
Is it something missing or is it because `onMetaData` is not really useful?

Best regards,
Thibault

**Comment by veovera** [comment]
Heya Thibault,

That is an interesting topic. The onMetaData marker is not always guaranteed to be there. That said it does not mean that we should not enhance onMetaData in conjunction with the latest rtmp updates. I will take a closer look. Out of curiosity how do you use the callback and do you use all the properties that the callback delivers?

**Comment by ThibaultBee** [comment]
Hi,

Sorry to say that I won't answer your question because I am focusing on the emitter part. I just noticed that `videocodecid` was missing when I tried to implement enhanced RTMP in my emitter library.
But one day, someone asks to add info in the `onMetaData`, so at least few people requires this marker.

**Comment by zenomt** [comment]
clarification of the above: the Enhanced RTMP spec defines the Video FourCC as a UI32 (rather than, say, UI8[4]), and FourCCs are usually interpreted as unsigned 32-bit integers out in the world anyway. since an onMetaData's `videocodecid` is already a Number, which has the precision to represent any UI32, and any codec ID that was a fourCC would be unambiguous with the existing codec ID space <= 15, it seems the most natural to just use the fourCC integer value as the `videocodecid`. any program that cares about the `videocodecid` in an onMetaData and knew about Enhanced RTMP would probably just have numeric definitions for the new enhanced codec IDs, maybe with a comment saying where those numbers came from.

**Comment by veovera** [comment]
> 
We probably will not overload the meaning of videocodecid and instead provide a new property which will identify fourCC via a string (i.e. same style as in fourCcList property, see NetConnection connect command)

**Comment by xia-chu** [comment]
> We probably will not overload the meaning of videocodecid and instead provide a new property which will identify fourCC via a string (i.e. same style as in fourCcList property, see NetConnection connect command)

Hi,  I'm sad to tell you that, at this moment,  we can not guess whether the video codec is  H264 or H265 from `obs`.
`Obs` send `metadata` with videocodecid of `7` when push a H265 rtmp stream:
![图片](https://github.com/veovera/enhanced-rtmp/assets/11495632/632e2f08-a58e-4f4b-9ed1-f2ffb5a4241b)


And `obs` did not set `fourCcList` in ` NetConnection connect command`:
![图片](https://github.com/veovera/enhanced-rtmp/assets/11495632/f3971999-cd43-4dba-8f81-efdef14fc4f0)


So when i am implementing `enhanced-RTMP`, i am very confused whether the stream is H264 or H265.
Rtmp metadata will becoming dispensable, it can even lead to ambiguity.

So I strongly recommend adding a field in metadata that indicates the correct codecid.

**Comment by veovera** [comment]
> > We probably will not overload the meaning of videocodecid and instead provide a new property which will identify fourCC via a string (i.e. same style as in fourCcList property, see NetConnection connect command)
> 
> Hi, I'm sad to tell you that, at this moment, we can not guess whether the video codec is H264 or H265 from `obs`. `Obs` send `metadata` with videocodecid of `7` when push a H265 rtmp stream: ![图片](https://user-images.githubusercontent.com/11495632/255112181-632e2f08-a58e-4f4b-9ed1-f2ffb5a4241b.png)
> 
> And `obs` did not set `fourCcList` in ` NetConnection connect command`: ![图片](https://user-images.githubusercontent.com/11495632/255112557-f3971999-cd43-4dba-8f81-efdef14fc4f0.png)
> 
> So when i am implementing `enhanced-RTMP`, i am very confused whether the stream is H264 or H265. Rtmp metadata will becoming dispensable, it can even lead to ambiguity.
> 
> Also, what videocodeid should be used for vp9 and av1?
> 
> So I strongly recommend adding a field in metadata that indicates the correct codecid.
> 
> Thanks, Best regards.

Yes, this is a known issue in the onMetadata callback and there is a plan to address it in the future. When parsing the bitstream are you able to look at the FourCC value which signals the current codec (see ExVideoTagHeader Description)?

**Comment by zenomt** [comment]
i'd like to draw attention to Discussion #19 "[Enhanced RTMP & FLV test/sample media, clients, and servers for validation and interop](https://github.com/veovera/enhanced-rtmp/discussions/19)". this is a case where attempting interoperation with different tools and media would surface issues like this earlier.

no matter what, the onMetaData `videocodecid` shouldn't be `7` if the codec is actually HEVC. i think the right thing to do for this specific case is to file a bug report for OBS.

**Comment by xia-chu** [comment]
> Yes, this is a known issue in the onMetadata callback and there is a plan to address it in the future. When parsing the bitstream are you able to look at the FourCC value which signals the current codec (see ExVideoTagHeader Description)?

Yes, we can discriminate H264 or H265 from VideoTagHeader by `FourCC` value.
But skip the metadata will caused some issues, such as delay increasement.
By the way, if metadata declare an incorrect vidoecodecid, this is not rigorous.

How about set `videocodecid` to zero and add a `FourCC` string value in metadata?

**Comment by veovera** [comment]
> How about set `videocodecid` to zero and add a `FourCC` string value in metadata?

Yes what is what we are thinking as well. Note: it is legal to not have videocodecid property in the onMetadata SCRIPTDATA Tag.  Which means we can to return a string representing the FourCC codec and not have videocodecid property present.

**Comment by veovera** [comment]
@xia-chu let us know if you have some flv content which would be ok to share to test this issue. thanks!

**Comment by xia-chu** [comment]
> Yes what is what we are thinking as well. Note: it is legal to not have videocodecid property in the onMetadata SCRIPTDATA Tag. Which means we can to return a string representing the FourCC codec and not have videocodecid property present.

If have no videocodecid in metadata, it may be mistaken for not having a video, i don't think this is appropriate.
An unsupported videocodecid is better than the absence of a videocodeid.

**Comment by xia-chu** [comment]
> let us know if you have some flv content which would be ok to share to test this issue. thanks!

I can generate a flv file with`metadata` which `videocodecid` id is 0 and `videoFourCC` is `hvc1` if you want.

**Comment by zenomt** [comment]
> I can generate a flv file `withmetadata` which `videocodecid` id is 0 and videoFourCC is hvc1 if you want.

i strongly recommend against having a `videocodecid` property that's `0`, since traditional codec ID 0 is "None". instead i reiterate my suggestion above: since Enhanced RTMP fourCC codec IDs are already numbers by definition (they're defined to be UI32 in the spec), that for purposes of the `videocodecid`, saying "video codec 1752589105 is HEVC" is the same effort as saying "video codec 7 is AVC".

in other words, the new fourCCs are already compatible with and a natural fit for the current way `videocodecid` works and is used.

**Comment by veovera** [comment]
> > let us know if you have some flv content which would be ok to share to test this issue. thanks!
> 
> I can generate a flv file with`metadata` which `videocodecid` id is 0 and `videoFourCC` is `hvc1` if you want.
Hi,

having a codec id of zero is problematic. In pre 2023 flv file spec audio/video codec ids are defined via 4 bit values. So zero either has a meaning or is reserved. For instance [audiocodecid of zero](https://rtmp.veriskope.com/pdf/video_file_format_spec_v10_1.pdf#page=76) (yes, we working on enhancing audio codecs as well...) stands for linear pcm, platform endian).

Given all the comments in this thread, as already suggested earlier we are now leaning towards encoding a FoucCC value in videocodecid property when dealing with enhanced-RTMP codecs. This should work since we only need 32 bits for a FourCC value and an AMF0 Numbers have more than 32 bits to encode an integer value. Further out we will mimic the same behavior on the audio side. Stay tuned for a description in the spec to formally align.

**Comment by xia-chu** [comment]
> we are now leaning towards encoding a FoucCC value in videocodecid property when dealing with enhanced-RTMP codecs

This is also acceptable， thanks for your work.

**Comment by veovera** [comment]
Closing the issue. We added new onMetaData section in the spec. videocodecid property MUST now be set to a FOURCC value (per enhanced-RTMP spec) when video codec is signaled via FourCC.
